src/Fcf/Data/List.hs

```diff
@@ -205,3 +208,78 @@ type instance Eval (Unzip as) = Eval (Foldr Cons2 '( '[], '[]) (Eval as))
 data Cons2 :: (a, b) -> ([a], [b]) -> Exp ([a], [b])
 type instance Eval (Cons2 '(a, b) '(as, bs)) = '(a ': as, b ': bs)
 
+
+-- Helper for the Take. This corresponds to the unsafeTake in the data list lib.
+data UtakeStop :: Nat -> [a] -> Exp [a]
+type instance Eval (UtakeStop 0 _)         = '[]
+type instance Eval (UtakeStop 1 (a ': as)) = '[a]
+
+-- Helper for the Take. This corresponds to the unsafeTake in the data list lib.
+data Utake :: Nat -> [a] -> Exp [a]
+-- As the following refuses to compile, we use UtakeStop-helper.
+-- type instance Eval (Utake 0 (a ': as)) = '[]
+-- type instance Eval (Utake 1 (a ': as)) = '[a]
+-- type instance Eval (Utake m (a ': as)) = Eval ('(:) a <$> (Utake (m TL.- 1) as))
```

Comment:
This definition between comments looks closer to the right one. It gets rejected because type instances are not allowed to overlap (`Utake 0 (a ': as)` and `Utake m (a ': as)`). Instead you can define `Utake` as a closed type family. Also, I'm not sure where you saw it, but there is no `unsafeTake` in `Data.List` (there is one for vectors and bytestrings, but it doesn't make as much sense for lists).

```
type family Take_ (n :: Nat) (xs :: [a]) :: [a] where
  Take_ 0 _ = '[]
  Take_ _ '[] = '[]
  Take_ n (x ': xs) = x ': Take_ n xs
```

Reply:
I was looking these at https://hackage.haskell.org/package/base-4.12.0.0/docs/src/GHC.List.html#take
and it looks like there are several possible definitions. I should have thought about this a bit: the report version is actually close to that closed familify def. I'll change it to use that one.

Reply:
Oh I didn't realize `base` had such optimized versions. Do use the simpler versions here!

Reply:
and now changed